### PR TITLE
Implement new concept with exercise: randomness

### LINKED ISSUE #2038: Implement new Concept Exercise: Randomness
## Getting Started

If you have not yet contributed to concept exercises before, this task requires some upfront reading to acquire the necessary background knowledge.

Here you can read about what Concept Exercises are and how they are structured:

- [What are Concept Exercises?](https://exercism.org/docs/building/product/concept-exercises)
- [Concept Exercises](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concepts](https://exercism.org/docs/building/tracks/concepts)

If you have not done so yet, it is probably also helpful to do a couple of "Learning Exercises" (this is how they are called on the site) yourself. You can also look at the code of an existing concept exercise like [`welcome-to-tech-palace`](https://github.com/exercism/go/tree/main/exercises/concept/welcome-to-tech-palace) for reference.

Also be aware of these general guidelines.

- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)

## Goal

The goal here is to create a new concept exercise to teach randomness (random number generation).
The concept needs to be written from scratch, the exercise can be ported from another track (see details below).

## Concepts

The following concept needs to be created. You can use the `introduction.md` file of the concept also as `introduction.md` file of the exercise. No need to create different content at this point. Additionally, if you want to save some time it is ok to not have an extensive `about.md` for now. It can also be mainly the `introduction.md` content, maybe with some additions you would like to make.

- `randomness`

## Learning Objectives

In the concepts the student should learn about the following topics and then practice them in the concept exercise.

- how to create pseudo random numbers with the `math/rand` package (float and int)
- what seeding is for and how/where to define the seed in the code
- that there are some helpers in the `math/rand` package line `Shuffle` and `Perm`
- when and how to create cryptographically secure random numbers with `crypto/rand` (does not have to be part of the exercise but would be good to mention in the concept to raise security awareness)

## Out of Scope

- details of pseudo random number generation in general
- discussing the different algorithms for random number generation
- details on entropy sources etc

## Prerequisites

- `numbers`

Other prerequisites should be added as needed to solve the specific exercise.

## Exercise Idea

[C# "Parsing Log Lines" Exercise](https://github.com/exercism/csharp/tree/main/exercises/concept/parsing-log-files) could serve as template.

In case you port that exercise, make sure to check all tasks make sense for Go and whether some additional task needs to be added to cover some Go specific content.

## Resources

Here some links that might be helpful as a starting point and/or for the links section of the concept:

- https://pkg.go.dev/math/rand
- https://pkg.go.dev/crypto/rand
- https://gobyexample.com/random-numbers

## How to proceed

1. Accept this issue by saying "I'd like to work on this" (no need to wait for a response, just go with it).
2. Use this issue to discuss any questions you have.
3. Create a PR and set "exercism/go" as reviewers when you are done. Additionally you can write in #maintaining-go that your PR is ready for review. Once you incorporated any critical feedback that the reviewer might give you and the PR is approved, it will be merged by a maintainer.

### PR DESCRIPTION
**Pull request by norbs57**
Fixes #2038
Implements concept randomness including an exercise
The exercise was partly ported from the roll-a-die exercise on the C# track

### CODE REVIEW

File: concepts/randomness/.meta/config.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "blurb": "math.rand can be used to generate pseudo-random integral or floating-point numbers.",
```
Comment (andrerfcsantos) [inline concepts/randomness/.meta/config.json:2]:
```suggestion
  "blurb": "Generate pseudo-random integer and floating-point numbers.",
```

File: concepts/randomness/about.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
+# Randomness Introduction
```
Comment (andrerfcsantos) [inline concepts/randomness/about.md:1]:
```suggestion
# About
```

File: concepts/randomness/about.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+Package [`math\rand`][mathrand] provides support for generating pseudo-random numbers.
```
Comment (andrerfcsantos) [inline concepts/randomness/about.md:3]:
```suggestion
Package [math\rand][mathrand] provides support for generating pseudo-random numbers.
```

Although the intention is good, my experience is that the block style might not make it obvious that the text is a link, so it's better to remove it.

Can you fix this in the other places too please?

File: concepts/randomness/about.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+By default, the package uses the number `1` as a seed.
+
+In order to get a different sequence of random numbers, we need to set a different seed.
+A simple approach is to seed with the current computer time.
+
+```go
+rand.Seed(time.Now().UnixNano())
+```
+
+Seeding must be done **before** any of the functions that generate random numbers.
```
Comment (andrerfcsantos) [inline concepts/randomness/about.md:40]:
```suggestion
Setting the seed must be done **before** calling any of the functions that generate random numbers. We usually only need to set the seed once in our program. From one seed we can generate many random numbers.
```

Additional clarification. Also apply this change in the other places where this is present.

File: concepts/randomness/about.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+265737 21561 640434
+
+708506 627509 363987
+```
+
+### Caution
+
+The output produced by package `math/rand` might be easily predictable regardless of how it is seeded.
+For random numbers suitable for security-sensitive work, you should use the [`crypto/rand`][cryptorand] package.
```
Comment (andrerfcsantos) [inline concepts/randomness/about.md:84]:
```suggestion
~~~~exercism/caution

The output produced by package `math/rand` might be easily predictable regardless of how it is seeded.
For random numbers suitable for security-sensitive work, you should use the [crypto/rand][cryptorand] package.
~~~~
```

We can take advantage of this special syntax exercism allows to highlight this block. To read more about this check [markdown](https://exercism.org/docs/building/markdown/markdown) in the docs.

Apply this change in the other places where this is present.

File: concepts/randomness/introduction.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Randomness Introduction
```
Comment (andrerfcsantos) [inline concepts/randomness/introduction.md:1]:
```suggestion
# Introduction
```

File: exercises/concept/animal-magic/.docs/introduction.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
+# Randomness Introduction
```
Comment (andrerfcsantos) [inline exercises/concept/animal-magic/.docs/introduction.md:1]:
```suggestion
# Introduction
```

File: concepts/randomness/about.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+### Advanced
+
+Instead of seeding with computer time, it is better to use `crypt/rand` functions to generate a seed for `math/rand`.
+For more information, see this [stackoverflow article][stackoverflow].
```
Comment (andrerfcsantos) [inline concepts/randomness/about.md:89]: I think there should be a brief explanation on why `crypt/rand` is "better". In what sense? Following the stack overflow link is also not clear for me what you mean. Maybe there's a better resource for it?

File: exercises/concept/animal-magic/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
+# Instructions
+
+Elaine is working on a new children's game that features animals and magic wands.
+It is time to code functions for rolling a die, generating random wand energy and shuffling a slice.
+
+## 1. Seed the random number generator.
+
+Implement a `SeedWithTime` function that seeds `math.rand` with the current computer time.
```
Comment (andrerfcsantos) [inline exercises/concept/animal-magic/.docs/instructions.md:8]:
```suggestion
Implement a `SeedWithTime` function that seeds the `math.rand` package with the current computer time.
```

File: concepts/randomness/introduction.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+```go
+rand.Seed(time.Now().UnixNano())
+```
+
+Seeding must be done **before** your program calls any of the functions that generate random numbers.
+
+### Warning
+
+If you don't call `rand.Seed`, then the package will produce the same sequence of random numbers each time your program runs!
```
Comment (andrerfcsantos) [inline concepts/randomness/introduction.md:44]:
```suggestion
~~~~exercism/caution

If you don't call `rand.Seed`, then the package will produce the same sequence of random numbers each time your program runs!
~~~~
```

**Review by andrerfcsantos — CHANGES_REQUESTED** [review]
Thanks for working on this. I really like the introduction to the subject you did, is short, straight to the point and the examples are relevant.

Left some minor comments.

File: concepts/randomness/about.md (hunk 3 above)
Comment (norbs57) [reply concepts/randomness/about.md:3]: Thanks - done.

File: concepts/randomness/about.md (hunk 4 above)
Comment (norbs57) [reply concepts/randomness/about.md:40]: Done :)

File: concepts/randomness/about.md (hunk 5 above)
Comment (norbs57) [reply concepts/randomness/about.md]: This caution is currently only in file `about.md`. Should it maybe also go in the concept introduction in the randomness folder?

File: concepts/randomness/about.md (hunk 8 above)
Comment (norbs57) [reply concepts/randomness/about.md:90]: In essence, using `time` as a seed means a lot of the bits in the seed will end up the same, especially when the program is run several times in short succession and computer time lacks precision. Using a `crypt/rand` generated seed gives more entropy, hence less predictable (and potentially less brute-forceable) random-numbers sequences generated by `math/rand`. So in principle, this looks like sound advice. Unfortunately, I have not been able to find a better resource to link to. I suggest we remove this section for now and keep it on a "under consideration" list for the future.

Comment (andrerfcsantos) [reply concepts/randomness/about.md]: I think we can remove it too. We can just mention that crypt/rand can be used for randomly secure cryptography applications and leave it at that.

Maybe explaining the difference between the packages would be more suitable in a future exercise where we talk about creating generators from different sources.

File: concepts/randomness/about.md (hunk 5 above)
Comment (andrerfcsantos) [reply concepts/randomness/about.md]: You are right, thought I've seen this somewhere else, disregard the advice then.

**Comment by norbs57** [comment]
You may have seen it on https://pkg.go.dev/math/rand

**Review by andrerfcsantos — APPROVED** [review]

**Pull request merged**